category_theory/functor/default.lean

```diff
@@ -0,0 +1,89 @@
+-- Copyright (c) 2017 Scott Morrison. All rights reserved.
+-- Released under Apache 2.0 license as described in the file LICENSE.
+-- Authors: Tim Baumann, Stephen Morgan, Scott Morrison
+
+import ..category
+
+open categories
+
+namespace categories.functor
+ 
+universes u₁ v₁ u₂ v₂ u₃ v₃
+
+structure Functor (C : Type u₁) [category.{u₁ v₁} C] (D : Type u₂) [category.{u₂ v₂} D] : Type (max u₁ v₁ u₂ v₂) :=
+  (onObjects     : C → D)
+  (onMorphisms   : Π {X Y : C}, (X ⟶ Y) → ((onObjects X) ⟶ (onObjects Y)))
+  (identities    : ∀ (X : C), onMorphisms (𝟙 X) = 𝟙 (onObjects X) . obviously)
+  (functoriality : ∀ {X Y Z : C} (f : X ⟶ Y) (g : Y ⟶ Z), onMorphisms (f ≫ g) = (onMorphisms f) ≫ (onMorphisms g) . obviously)
+
+make_lemma Functor.identities
+make_lemma Functor.functoriality
+attribute [simp,ematch] Functor.functoriality_lemma Functor.identities_lemma
+
+infixr ` +> `:70 := Functor.onObjects
+infixr ` &> `:70 := Functor.onMorphisms -- switch to ▹?
+infixr ` ↝ `:70 := Functor              -- type as \lea 
+
+definition IdentityFunctor (C : Type u₁) [category.{u₁ v₁} C] : C ↝ C := 
+{ onObjects     := id,
+  onMorphisms   := λ _ _ f, f,
+  identities    := begin 
+                     -- `obviously'` says:
+                     intros,
+                     refl 
+                   end,
+  functoriality := begin
+                     -- `obviously'` says:
+                     intros,
+                     refl
+                   end }
+
+instance (C) [category C] : has_one (C ↝ C) :=
+{ one := IdentityFunctor C }
+
+section
+variable {C : Type u₁}
+variable [category.{u₁ v₁} C]
+
+@[simp] lemma IdentityFunctor.onObjects   {C : Type u₁} [category.{u₁ v₁} C] (X : C)                 : (IdentityFunctor C) +> X = X := by refl
```

Comment:
The colon doesn't algin with the line below, but it looks like you want it to.

Reply:
It seems there are duplicate variables `{C : Type u₁} [category.{u₁ v₁} C]` here?

Reply:
... and below here, as well.

Reply:
I couldn't get it to align exactly (slight variations in character widths), so I've removed all the spurious spaces.

@spl, I don't think there was any actual duplication, because of sectioning commands, but in any case I've tidied the variables up.